File: go/prime_sieve.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func primeSieve(n int) int {
	isPrime := make([]bool, n+1)
	for i := 0; i <= n; i++ {
		isPrime[i] = true
	}
	isPrime[0] = false
	isPrime[1] = false
	
	for i := 2; i*i <= n; i++ {
		if isPrime[i] {
			for j := i * i; j <= n; j += i {
				isPrime[j] = false
			}
		}
	}
	
	count := 0
	for i := 2; i <= n; i++ {
		if isPrime[i] {
			count++
		}
	}
	
	return count
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: prime_sieve <n>")
		return
	}

	n, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Println("Invalid number:", os.Args[1])
		return
	}

	start := time.Now()
	count := primeSieve(n)
	elapsed := time.Since(start)

	timeMs := elapsed.Milliseconds()

	fmt.Printf("Go: prime_sieve(%d) = %d\n", n, count)
	fmt.Printf("Time: %dms\n", timeMs)
}

File: c/prime_sieve.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

int prime_sieve(int n) {
    int* is_prime = malloc((n + 1) * sizeof(int));
    if (!is_prime) return -1;
    
    // Initialize all as prime
    for (int i = 0; i <= n; i++) {
        is_prime[i] = 1;
    }
    is_prime[0] = is_prime[1] = 0;
    
    for (int i = 2; i * i <= n; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= n; j += i) {
                is_prime[j] = 0;
            }
        }
    }
    
    int count = 0;
    for (int i = 2; i <= n; i++) {
        if (is_prime[i]) count++;
    }
    
    free(is_prime);
    return count;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Usage: prime_sieve <n>\n");
        return 1;
    }
    
    int n = atoi(argv[1]);
    
    clock_t start = clock();
    int count = prime_sieve(n);
    clock_t end = clock();
    
    double time_ms = ((double)(end - start) / CLOCKS_PER_SEC) * 1000;
    
    printf("C: prime_sieve(%d) = %d\n", n, count);
    printf("Time: %.0fms\n", time_ms);
    
    return 0;
}